<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Eff3ctidor</title>

    <link rel="shortcut icon" type="image/x-icon" href="/pics/logo.gif" />

    <meta charset="utf-8">
    <!--<meta name="viewport" content="width=device-width, initial-scale=1">-->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <link rel="stylesheet" href="https://gitcdn.github.io/bootstrap-toggle/2.2.2/css/bootstrap-toggle.min.css">
    
    <!--files for the select tag-->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/css/bootstrap-select.min.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/bootstrap-select.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/i18n/defaults-*.min.js"></script>

    <!--files for the checkbox tag (toggle)-->
    <script src="https://gitcdn.github.io/bootstrap-toggle/2.2.2/js/bootstrap-toggle.min.js"></script>

    <link rel="stylesheet" href="css/general.css">
    <script src="handle_FAQ.js"></script>
</head>
<body>
    <nav role="navigation" class="navbar navbar-fixed-top">
        <div class="jumbotron" id="jumbo">
            <div class="container">            
                <div class="row" id="title-row" align="center">
                    <div class="col-md-1">
                    </div>
                    <div class="col-md-10" align="center">
                        <!--<span id="server-title">Effectidor</span><br>-->
                        <img src="pics/logo.gif" id="nav_bar_image" style="height: 120px;"><br>
                        <span id="sub-title">A machine learning based type 3 effectors predictor</span>
                        <br><br>
                    </div>
                </div>
            </div>       
        </div>
        <div class="jumbotron" id="nav_container">
                <div class="row">
                    <div style="display:table; margin:0 auto;">
                        <ul class="nav navbar-nav" id="menu-nav">
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="index.html">Home</a>
                            </li>
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="FAQ.html">FAQ</a>
                            </li>
                            <li class="active" style="display:table-cell; padding: 0px 10px;">
                                <a href="data.html">Data</a>
                            </li>
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="credits.html">Credits</a>
                            </li>
                            <li style="float: right">
                                <a href="https://english.tau.ac.il/" id="tau-link" target="_blank">
                                    <img src=pics/TAU_LOGO.png id="tau-logo">
                                </a>
                            </li>
                        </ul>
                    </div>
                </div>
        </div>
    </nav>
    <div id="behind-nav-bar-results"></div>
    <div class="container">
    <div class="col-md-1"></div>
    <div class="col-md-10">
        <p align="justify">
        Effectidor identifies T3SS and flagella proteins in the input genome(s) based on internal datasets of the different T3SS subtypes. These proteins are excluded from the analysis of Effectidor and are reported separately.<br>
        These datasets are available for downloading:<br>
        <b><a href="T3SS_data/Chlamydiales.fasta" target="_blank">Chlamydiales</a></b><br>
        <b><a href="T3SS_data/Ysc.fasta" target="_blank">Ysc</a></b><br>
        <b><a href="T3SS_data/SPI-1.fasta" target="_blank">SPI-1</a></b><br>
        <b><a href="T3SS_data/SPI-2.fasta" target="_blank">SPI-2</a></b><br>
        <b><a href="T3SS_data/Hrp1.fasta" target="_blank">Hrp1</a></b><br>
        <b><a href="T3SS_data/Hrp2.fasta" target="_blank">Hrp2</a></b><br>
        <b><a href="T3SS_data/Rhc.fasta" target="_blank">Rhc</a></b><br>
        <b><a href="T3SS_data/Flagellar.fasta" target="_blank">Flagella</a></b><br><br>
        </p>
    </div></div>
        <div class="container">
        <div class="col-md-1"></div>
    <div class="col-md-10">
        <p align="justify">
        Effectidor uses a type-III effectors dataset for homology search as well as for composing the positive group if no effectors file is supplied.<br>
        The different versions of this dataset are available for downloading:<br><br>
        <b><a href="T3Es_data/T3Es9.faa" target="_blank">T3Es version 1.09</a></b> (February 2025) - <b>removed False T3Es:</b>
            B7I4B5, D4XM62, D0SEN9, D0BWB6, C6RTB7, C0VHV6, B3PEW2, A4VIM7 (distant homologs of HopPmaJ from bacteria without a T3SS),
            P0A3W8 (similar to a T4E from a bacterium without a T3SS), D1MWR4 (transposase), Q89N83 (NodN),
            Q9Z7E2 (toxin-antitoxin system protein) and Q9Z8G9 (peptidoglycan editing factor PgeF) [<a href="https://www.nature.com/articles/ng0499_385" target="_blank">REF</a>],
            E3DFP4 (potential T3SS chaperone) [<a href="https://journals.asm.org/doi/10.1128/jb.01246-10" target="_blank">REF</a>],
            YP_237108.1 (T6E Tae4), Q7BEJ1 (toxin-antitoxin system protein), ExoT, B8GTM1 (from a bacterium lacking the T3SS),
            C5ALL7 (phage tail tip lysozyme), O84436, O84342, and O84854 (tested and were not secreted) [<a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC11071493/" target="_blank">REF</a>],
            Q9REZ7 (needle tip) [<a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1365-2958.2008.06543.x" target="_blank">REF</a>],
            Q84CS5 (part of the T3SS OM ring), Q9Z6N8 and Q9Z8L3 (T3S chaperones) [<a href="https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1004498" target="_blank">REF</a>],
            O84944 (T3SS chaperone sseA) [<a href="https://onlinelibrary.wiley.com/doi/full/10.1046/j.1365-2958.2003.03373.x" target="_blank">REF</a>],
            A1JU78 and P0C2N4 (component of the T3SS, YscX [<a href="https://journals.asm.org/doi/10.1128/jb.181.2.675-680.1999">REF</a>]).
            <br>
            <b>added validated T3Es:</b>
            APS58_1000* (APS58_1000star) [<a href="https://bsppjournals.onlinelibrary.wiley.com/doi/full/10.1111/mpp.12877" target="_blank">REF</a>],
            IncA and CPj0517/CP0236 [<a href="https://doi.org/10.1093/nar/28.12.2311" target="_blank">REF1</a>,<a href="https://doi.org/10.1128/JB.00511-20" target="_blank">REF2</a>],
            CT144, CT016, and CT622 [<a href="https://www.biorxiv.org/content/10.1101/2024.04.25.587017v1.full.pdf" target="_blank">REF</a>],
            copN [<a href="https://onlinelibrary.wiley.com/doi/10.1046/j.1365-2958.2000.02212.x" target="_blank">REF</a>],
            CopB [<a href="https://journals.asm.org/doi/full/10.1128/jb.187.18.6466-6478.2005" target="_blank">REF</a>],
            CT058, CT195, CT226, CT228, CT249, CT324, CT358, CT383, and CT440 [<a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-12-109" target="_blank">REF</a>],
            CT671, CT579, CT860, CT_652.1, CT_606.1, and incE [<a href="https://onlinelibrary.wiley.com/doi/epdf/10.1111/j.1365-2958.2005.04647.x" target="_blank">REF</a>],
            CTL0404 [<a href="https://journals.asm.org/doi/pdf/10.1128/iai.00491-22" target="_blank">REF</a>],
            incC [<a href="https://onlinelibrary.wiley.com/doi/10.1046/j.1365-2958.2003.03462.x" target="_blank">REF</a>],
            pemB and pemA [<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4453518/" target="_blank">REF</a>],
            avrRpm2 and HOPA1 [<a href="https://apsjournals.apsnet.org/doi/epdf/10.1094/MPMI-07-13-0206-R" target="_blank">REF</a>],
            SrfJ [<a href="https://journals.asm.org/doi/10.1128/jb.00173-12" target="_blank">REF</a>],
            OspZ [<a href="https://doi.org/10.1074/jbc.m116.734079" target="_blank">REF</a>],
            SG0764 [<a href="https://doi.org/10.1128/mbio.00243-11" target="_blank">REF</a>],
            XopXccLR [<a href="https://apsjournals.apsnet.org/doi/epdf/10.1094/MPMI-22-11-1401" target="_blank">REF</a>],
            AvrXccC [<a href="https://bsppjournals.onlinelibrary.wiley.com/doi/pdfdirect/10.1111/j.1364-3703.2007.00409.x" target="_blank">REF</a>],
            HrpE3 [<a href="https://bsppjournals.onlinelibrary.wiley.com/doi/10.1111/mpp.12039" target="_blank">REF</a>].
            <br>

        <br>Older versions:<br>
        <b><a href="T3Es_data/T3Es8.faa" target="_blank">T3Es version 1.08</a></b> (August 2023) - removed False T3E - HrpA.<br>
        <b><a href="T3Es_data/T3Es7.faa" target="_blank">T3Es version 1.07</a></b> (July 2023) - added newly validated T3E of <i>Ralstonia</i>, RipBT [<a href="https://bsppjournals.onlinelibrary.wiley.com/doi/10.1111/mpp.13342" target="_blank">REF</a>].<br>
        <b><a href="T3Es_data/T3Es6.faa" target="_blank">T3Es version 1.06</a></b> (June 2023) - added newly validated T3Es from <i>Xanthomonas</i>: XopBB, XopBC, XopBD, XopBE, XopBF, XopBG [<a href="https://www.frontiersin.org/articles/10.3389/fpls.2023.1155341/full" target="_blank">REF</a>].<br>
        <b><a href="T3Es_data/T3Es5.faa" target="_blank">T3Es version 1.05</a></b> (January 2023) - added validated T3Es from <i>Xanthomonas</i>, <i>Pseudomonas syringea</i>, <i>Ralstonia</i>, <i>Acidovorax</i>, and <i>Panotea</i> sp.<br>
        <b><a href="T3Es_data/T3Es4.faa" target="_blank">T3Es version 1.04</a></b> (November 2022) - added the following validated T3Es: AvrA, AvrB6, AvrBs2, AvrBs4, AvrXa3, AvrXa7, BapC, CigR, HopAF1, HopAK1, HopAO1, HopD1, HopO1-1, HopQ1-1, HopZ1, HrpE3, NopF, NopT, PopW, PopW, PthA, PthA2, PthA3, PthA4, PthB, PthC, PthG, RipAL, RipB, RipTPS, SrfJ, XAC3090, XOO3150, XopA, XopAC, XopAF, XopAG, XopAM, XopAO, XopAP, XopAR, XopAS, XopAU, XopAW, XopC, XopD, XopE1, XopE2, XopF1, XopF2, XopG, XopH, XopI, XopJ2, XopJ4, XopJ5, XopK, XopM, XopN, XopP, XopQ, XopS, XopT, XopV, XopX, XopZ, XopZ1.<br>
        <b><a href="T3Es_data/T3Es3.faa" target="_blank">T3Es version 1.03</a></b> (September 2022) - removed false T3E E2MEI8 and added missing <i>Ralstonia</i> T3Es and the following T3Es: XopZ2, XopJ4, XopJ2, XopE5, XopAS, XopAQ, XopAL2, XopAF, AvrXccA1.<br>
        <b><a href="T3Es_data/T3Es.faa" target="_blank">T3Es version 1.01</a></b> (October 2021)</p>
    </div>
    </div>
    <div id="behind-footer"></div>
    <nav role="navigation" class="navbar-inverse navbar-fixed-bottom ng-binding" style="color: #cccccc; text-align: center; font-family: Calibri;">
        <h4>
                Research Site | <a href="https://www.tau.ac.il/~talp/" target="_blank">Pupko Group</a>
        </h4>
    </nav>
</body>
</html>