<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Eff3ctidor</title>

    <link rel="shortcut icon" type="image/x-icon" href="/pics/logo.gif" />

    <meta charset="utf-8">
    <script src="jquery-3.7.1.min.js"></script>
    <link href="bootstrap-5.3.3-dist/css/bootstrap.min.css" rel="stylesheet">
    <script src="bootstrap-5.3.3-dist/js/bootstrap.min.js"></script>


    <link rel="stylesheet" href="css/general.css">
    <script src="handle_FAQ.js"></script>
</head>
<body>
    <div class="row" id="jumbo">
            <div class="col-md-1">
            </div>
            <div class="col-md-10" align="center">
                 <img src="pics/logo.gif" id="nav_bar_image" style="height: 120px;"><br>
                 <span id="sub-title">A machine-learning based type III effectors predictor</span><br>
				 <span style="font-size:22px">Prof. Tal Pupko Lab - The Shmunis School of Biomedicine and Cancer Research</span>
                 <br><br>
            </div>
   </div>
   <nav class="navbar navbar-expand-lg bg-body-tertiary" role="navigation">
	    <div class="container-fluid" id="nav_container">
                    <div style="display:table; margin:0 auto;">
                        <ul class="nav navbar-nav"  id="menu-nav">
                            <li class="nav-item" style="padding: 0px 20px;"><a href="index.html">Home</a></li>
                            <li class="nav-item" style="padding: 0px 20px;"><a href="tutorial.html">Tutorials</a></li>
                            <li class="nav-item active" style="padding: 0px 20px;"><a href="FAQ.html">FAQ</a></li>
                            <li class="nav-item" style="padding: 0px 20px;"><a href="data.html">Data</a></li>
                            <li class="nav-item" style="padding: 0px 20px;"><a href="credits.html">Credits</a></li>
                            <li class="nav-item" style="padding: 0px 20px;"><a href="https://effectidor.tau.ac.il/results/examples/Example_run/output.html">Example run</a></li>
                            <li class="nav-item" style="padding: 0px 20px;float: right"><a href="https://english.tau.ac.il/" id="tau-link" target="_blank"><img src=pics/TAU_LOGO.png id="tau-logo"></a></li>
                        </ul>
                    </div>
           </div>
        </nav>     <!-- Static navbar -->
        
    <div class="container">
        <div class="col-md-1"></div>
    <div class="col-md-10">
    <br>
    <h3 onclick="Open_output()" style="cursor:pointer;">What is Effectidor's output?</h3><br>
    <div id="outputQ" style="display:none;">
    <p>Effectidor produces both tabular and graphical outputs.
    In the tabular output there are:<br><br><b>On screen tables</b>
    <ol>
        <li>The positive set used to train the classifier. These samples are sorted according to their likelihood to encode T3Es.</li>
        <li>Top 10 predictions. These predictions are the highest scoring putative T3Es, identified by Effectidor. Of note, the scores of these predictions should be taken into account in comparison to the scores of the known T3Es given in the previous table.</li>
    </ol></p>
    <p><b>Downloadable files</b>
    <ol>
        <li>Predictions file (<a href="tutorial_videos/output_1_predictions.mp4" target="_blank">tutorial video</a>). All the OGs in the genomes (excluding OGs that represent T3SS and flagella components) are sorted according to their score, reflecting their likelihood to encode T3Es. The label of each OG is given, as well as ORFs identity from each genome and protein annotation (if available in the input data).</li>
        <li>Feature importance file (<a href="tutorial_videos/output_3_features.mp4" target="_blank">tutorial video</a>). This table lists the features used for the ML, sorted by their contribution to the learning process, and their relative importance.</li>
        <li>Report on type 3 secretion system and flagella proteins found in the input genomes (<a href="tutorial_videos/output_2_T3SS_and_flagella.mp4" target="_blank">tutorial video</a>). These proteins are identified on the basis of sequence similarity to our T3SS datasets (see data). They are excluded from the ML process and therefore do not get a score, but are given as an extra output. To each subsystem found in the data all the components and their loci in each genome are listed.</li>
        <li>Raw features file (<a href="tutorial_videos/output_3_features.mp4" target="_blank">tutorial video</a>). This file contains all the features and their values for each OG in the data. These values were the input to the ML pipeline.</li>
    </ol></p>
    <br><p>
    The graphical output includes analysis of the features. The 10 most contributing features will be compared between the two classes - effectors and non-effectors, in violin plots.<br>
        Additional graphical output include presence/absence maps of the T3Es and T3SS and flagella components found in the genomes.
    <br><br>You can see an example for the output in the running example <a href="/results/examples/Example_run/output.html" target="_blank">here</a>.
    </p><br>
    </div>
    <h3 onclick="Open_interpretation()" style="cursor:pointer;">How do I interpret the results?</h3><br>
    <div id="interpretationQ" style="display:none;">
    <p>The main results, the predictions file, contains the OGs sorted by their likelihood to encode T3Es.
    <br>To identify candidates to be novel T3Es, search for OGs with a high score at the top rows of the table.
    <br>If you used an input Effectors file, look for samples that are not already known to be effectors. These samples will have a question mark (?) in the column "is_effector".
    <br>If you did not use an input Effectors file, the samples marked with "yes" in the "is_effector" column are homologous to known T3Es, and the highly scored samples marked with a question mark (?) are potentially novel T3Es.
    <br>For your convenience, the highly scored samples are highlighted with dark green that becomes lighter as the score decreases.
    <br>For each OG the annotation (if exists in the input file) is provided, and its members in each genome are specified.
    <br><br>In general, an OG with a high score and an ambiguous annotation is a good candidate to be a novel T3E.
        <br>a tutorial video for the predictions output is available <a href="tutorial_videos/output_1_predictions.mp4">here</a>.
    </p>
    <br>
    </div>
    <h3 onclick="Open_evaluation()" style="cursor:pointer;">How is Effectidor evaluated and where can I see its predicting accuracy?</h3><br>
    <div id="evaluationQ" style="display:none;">
    <p>In the learning process Effectidor randomly leaves 20% of the labeled data (effectors and non-effectors) aside, as a test set. Each classification algorithm is trained on the remaining 80% training data, using cross validation, and then evaluated on the untouched 20% test set.
    <br>The measurement used to evaluate Effectidor's performance is the Area Under the Precision Recall Curve (AUPRC). The closer it is to 1, the more accurate the classifier is. You can see the resulting AUPRC achieved on the test set in the downloadable predictions file.
    </p><br>
    </div>
    <h3 onclick="Open_input()" style="cursor:pointer;">What input does Effectidor require and what is it used for?</h3><br>
    <div id="inputQ" style="display:none;">
    <p>
    Effectidor has one obligatory input - an <b>ORFs file</b>.<br> This is a FASTA file including all the genome ORFs. See instructions for downloading this file <a href="instructions.html" target="_blank">here</a> and tutorial video for a single genome <a href="tutorial_videos/input_1_ORFs_single_genome.mp4" target="_blank">here</a>, and for pan-genome <a href="tutorial_videos/input_2_ORFs_multiple_genomes.mp4" target="_blank">here</a>.
    <br>Some of the ORFs in this file (effectors and non-effectors) will be used to train the machine-learning algorithms, and based on the trained classifier, the main output - prediction for each ORF - will be performed. 
    <br>In addition, it is possible to supply a known <b>Effectors file</b>, as appearing in the ORFs file. Alternatively, a homology search against an internal effectors' dataset will be performed to constitute the "known effectors" ORFs for the learning process. You can see a tutorial video for creating this input <a href="tutorial_videos/input_3_input_effectors.mp4" target="_blank">here</a>
    <br><br> In the advanced options you can supply data that will result in additional features to feed the machine-learning and improve the predictions. These data include:</p>
    <ol>
        <li><b>Host proteome</b> archive. Protein <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> files with the proteome of a known host of the studied bacterium. Multiple files can be included. All these files should be compressed in a single <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip</a> archive.
        <br>This input will be used for homology searches. As effectors interact with host proteins for their function, we expect them to have eukaryotic domains, that will be recognized in this homology search.
            <br> a tutorial video for this input is available <a href="tutorial_videos/input_6_host_proteome.mp4" target="_blank">here</a>.
        </li>
        <li>Archive of <b>proteomes of closely related bacteria without T3SSs</b>. This archive may contain several proteome records, each in a separate <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> file. These FASTA files should be compressed in a single <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip</a> archive.
        A homology search will be performed against each of these proteomes. As these bacteria are closely related to the studied bacterium, the vast majority of the proteins in the studied bacterium are expected to have an ortholog in these proteomes. Nevertheless, since they do not encode a T3SS, effectors are not expected to have orthologs in these proteomes. Thus, these features are usually very informative for the machine-learning.
            <br> a tutorial video for this input is available <a href="tutorial_videos/input_7_bacteria_without_T3SS.mp4" target="_blank">here</a>.
        </li>
        <li>
        <b><a href="http://gmod.org/wiki/GFF3" target="_blank">GFF3</a> file(s)</b>. These files will be used to compute genome organization features.
            <br> a tutorial video for this input is available <a href="tutorial_videos/input_8_gff3_full_genome_regulatory_elements.mp4" target="_blank">here</a>.
        </li>
        <li>
        <b>Full genome <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> files</b>. The full genome will be used to search for regulatory elements in the promoter region of each ORF. Speciffically, we allow searching for the following motifs: PIP-box, relevant for <i>Xanthomonas</i>, <i>Ralstonia</i>, and <i>Acidovorax</i>. hrp-box, relevant for <i>Pseudomonas syringae</i> and plant pathogens of the Enterobacteria family. mxiE-box, relevant for <i>Shigella</i>. exs-box, relevant for <i>Pseudomonas aeruginosa</i>. tts-box which is relevant for rhizobia.
            <br> a tutorial video for this input is available <a href="tutorial_videos/input_8_gff3_full_genome_regulatory_elements.mp4" target="_blank">here</a>.
        </li>
    </ol>
    </div>
    <h3 onclick="Open_running_time()" style="cursor:pointer;">What is the expected running time of Effectidor?</h3><br>
    <div id="running_timeQ" style="display:none;">
    <p>The running time depends on many factors, among them the input you supply and the load in our servers.
    <br>It can take a few minutes and up to several hours. In extreme cases, for example, when analyzing many genomes simultaneously, it can take several days. We will email you a link to the results upon submission, and upon completion.
    </p><br>
    </div>
    <h3 onclick="Open_saving_time()" style="cursor:pointer;">For how long will my results be saved in the servers?</h3><br>
    <div id="saving_timeQ" style="display:none;">
    <p>The results will be saved in our servers for 3 months.
    <br>After 3 months they will be permanently deleted from our servers.</p><br>
    </div>
    <h3 onclick="Open_email()" style="cursor:pointer;">Why do you need my email?</h3><br>
    <div id="emailQ" style="display:none;">
    <p>We will use it to send you a link to your results.<br>
    </div>
    <h3 onclick="Open_genomes()" style="cursor:pointer;">Can I use Effectidor to run the analysis on several genomes simultaneously?</h3><br>
    <div id="genomesQ" style="display:none;">
    <p>Yes! Effectidor V2 supports pan-genome analysis. The training and prediction are done on ortholog groups (OGs).</p>
    </div>
    </div></div>
    <div id="behind-footer"></div>
    </container>
    <div class="row" style="color: #cccccc; text-align: center; font-family: Calibri;" id="nav_container">
        <h4>
                Research Site | <a href="https://www.tau.ac.il/~talp/" target="_blank" style="color: #0d6efd;">Pupko Group</a>
        </h4>
    </div>
</body>
</html>
