<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Eff3ctidor</title>

    <link rel="shortcut icon" type="image/x-icon" href="/pics/logo.gif" />

    <meta charset="utf-8">
    <!--<meta name="viewport" content="width=device-width, initial-scale=1">-->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <link rel="stylesheet" href="https://gitcdn.github.io/bootstrap-toggle/2.2.2/css/bootstrap-toggle.min.css">
    
    <!--files for the select tag-->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/css/bootstrap-select.min.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/bootstrap-select.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/i18n/defaults-*.min.js"></script>

    <!--files for the checkbox tag (toggle)-->
    <script src="https://gitcdn.github.io/bootstrap-toggle/2.2.2/js/bootstrap-toggle.min.js"></script>

    <link rel="stylesheet" href="css/general.css">
    <script src="form_handler.js"></script>
</head>
<body>
    <nav role="navigation" class="navbar navbar-fixed-top">
        <div class="jumbotron" id="jumbo">
            <div class="container">            
                <div class="row" id="title-row" align="center">
                    <div class="col-md-1">
                    </div>
                    <div class="col-md-10" align="center">
                        <!--<span id="server-title">Effectidor</span><br>-->
                        <img src="pics/logo.gif" id="nav_bar_image" style="height: 120px;"><br>
                        <span id="sub-title">A machine-learning based type III effectors predictor</span>
                        <br><br>
                    </div>
                </div>
            </div>       
        </div>
        <div class="jumbotron" id="nav_container">
                <div class="row">
                    <div style="display:table; margin:0 auto;">
                        <ul class="nav navbar-nav" id="menu-nav">
                            <li class="active" style="display:table-cell; padding: 0px 10px;">
                                <a href="index.html">Home</a>
                            </li>
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="FAQ.html">FAQ</a>
                            </li>
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="data.html">Data</a>
                            </li>
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="credits.html">Credits</a>
                            </li>
                            <li style="float: right">
                                <a href="https://english.tau.ac.il/" id="tau-link" target="_blank">
                                    <img src=pics/TAU_LOGO.png id="tau-logo">
                                </a>
                            </li>
                        </ul>
                    </div>
                </div>
        </div>
    </nav>
    <div id="behind-nav-bar-results"></div>
    <div class="container">
    <div class="col-md-1"></div>
    <div class="col-md-10">
        <!--<div class="container" style="width: 100%">-->
        <br><br>
            <div class="form-group" id="short_description">
                <p align="justify">Effectidor is aimed to predict type III effectors in input bacterial genomes.
                Graphical and tabular results are displayed as output.
                Example output can be found <b><a href="/Example_run/output.html" target="_blank">here</a></b>.</p>
            </div>
            <br>
        </div>
        <!--<div class="col-md-1">-->
        </div>
    <!--</div>-->
    <div class="container">
    <div class="col-md-1"></div>
    <div class="col-md-11">
    <!--<div class="container" style="width: 100%">-->
        <!--<div class="col-md-1">-->

        <form ID="EFFECTIDOR_form" method="post" action="https://effectidor.tau.ac.il/cgi-bin/webserver_cgi.py" ENCTYPE="multipart/form-data" onsubmit="return validate_read_files()">
            <div>
                <label for="ORFs">Select an input ORFs file: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info1">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                <label class="btn btn-default btn-file">
                <input type="file" id="ORFs" name="ORFs" required>
                </label>
            </div>
            <br>
            <div>
                <label for="effectors">Select an input effectors file (optional): <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info2">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                <label class="btn btn-default btn-file">
                <input type="file" id="effectors" name="effectors" onchange="openOption()">
                </label>
            </div>
            <br>
            <div class="form-group row" id="homology_effectors" style="display:none;">
                <div class="col-md-10">
                    <label for="homology_search">Do you want us to add putative effectors to your effectors' data, that will be used together with your data as positive samples, based on homology to our effectors dataset? <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info3">
                            <span class="glyphicon glyphicon-question-sign"></span>
                    </button></label>
                    <label class="radio-inline"><input type="radio" name="homology_search" id="homology_search_yes" value="yes">Yes</label>
                <label class="radio-inline"><input type="radio" name="homology_search" id="homology_search_no" value="no" checked>No</label>
                </div>
            </div>
            <br>
            <div class="form-group row">
            <div class="col-md-4">
                <label for="email">E-mail:</label><br>
                <input type="email" id="email" name="email" class="form-control" placeholder="We'll send you a link to the results" required>
            </div>
            <div class="col-md-4">
                <label for="job_title">Job title:</label>
                <input type="text" id="job_title" name="job_title" class="form-control">
            </div>
            </div>
            <div>
                <label for="confirm_email"></label>
                <input type="hidden" id="confirm_email" name="confirm_email">
            </div>
            <button type="button" class="btn btn-warning" id="advanced_button" onclick="openAdvanced()"><span class="glyphicon glyphicon-cog"></span>&nbsp; Advanced options</button>
                <div id="advanced" style="display:none;">
                    <div>
                        <label for="translocation_signal">Would you like to include a prediction of the secretion signal in your analysis?
                            <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info12">
                                    <span class="glyphicon glyphicon-question-sign"></span>
                            </button></label><br>
                            <label class="radio-inline"><input type="radio" name="translocation_signal" id="translocation_signal_yes" value="yes">Yes</label>
                        <label class="radio-inline"><input type="radio" name="translocation_signal" id="translocation_signal_no" value="no" checked>No</label>
                    </div>
                    <br>
                    <div>
                        <label for="signalp">Would you like to include a prediction of non type-III secretion signal (using SignalP6)?
                            <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info13">
                                    <span class="glyphicon glyphicon-question-sign"></span>
                            </button></label><br>
                            <label class="radio-inline"><input type="radio" name="signalp" id="signalp" value="yes">Yes</label>
                        <label class="radio-inline"><input type="radio" name="signalp" id="signalp" value="no" checked>No</label>
                    </div>
                    <br>
                    <div class="form-group row">
                        <div class="col-md-8">
                            <label for="identity_cutoff">Minimal identity percentage of orthologs and paralogs:
                            <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info4">
                                <span class="glyphicon glyphicon-question-sign"></span>
                            </button></label>
                            <input type="number" class="form-control" min="0" max="100" step="any" id="identity_cutoff" name="identity_cutoff" value="50.0" required>
                        </div>
                    </div>
                    <div class="form-group row">
                        <div class="col-md-8">
                            <label for="coverage_cutoff">Minimal coverage percentage of orthologs and paralogs:
                            <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info5">
                                <span class="glyphicon glyphicon-question-sign"></span>
                            </button></label>
                            <input type="number" class="form-control" min="0" max="100" step="any" id="coverage_cutoff" name="coverage_cutoff" value="60.0" required>
                        </div>
                    </div>
                    <div class="form-group row">
                        <div class="col-md-10">
                            <label for="T3Es">Upload effectors file for homology search: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info6">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                            <label class="btn btn-default btn-file">
                            <input type="file" id="T3Es" name="T3Es">
                            </label>
                        </div>
                    </div>
                    <div class="form-group row">
                        <div class="col-md-10">
                            <label for="host">Upload a host proteome archive: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info7">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                            <label class="btn btn-default btn-file">
                            <input type="file" id="host" name="host">
                            </label>
                        </div>
                    </div>
                    <div class="form-group row">
                        <div class="col-md-10">
                            <label for="no_T3SS">Upload an archive with proteome records of closely related bacteria without T3SS: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info8">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                            <label class="btn btn-default btn-file">
                            <input type="file" id="no_T3SS" name="no_T3SS">
                            </label>
                        </div>
                    </div>
                    
                    <!--<div class="form-group row">-->
                    <!--    <div class="col-md-8">-->
                    <!--        <label for="full_genome">Is the genome fully assembled? (In one, circular, contig) <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info6">-->
                    <!--                        <span class="glyphicon glyphicon-question-sign"></span>-->
                    <!--                    </button></label>-->
                    <!--        <label class="radio-inline"><input type="radio" name="full_genome" id="full_genome_yes" value="yes" onclick="ShowHideDiv()">Yes</label>-->
                    <!--        <label class="radio-inline"><input type="radio" name="full_genome" id="full_genome_no" value="no" checked onclick="ShowHideDiv()">No</label>-->
                    <!--    </div>-->
                    <!--</div>-->
                    
                    
                        <div>
                            <div class="form-group row">
                                <div class="col-md-10">
                                    <label for="gff">Upload a <a href="http://gmod.org/wiki/GFF3" target="_blank">GFF3</a> file: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info9">
                                    <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                                    <label class="btn btn-default btn-file">
                                    <input type="file" id="gff" name="gff" onchange="ShowHideDiv()">
                                    </label>
                                </div>
                            </div>
                        </div>
                        <br><br>
                    <div id="promoters" style="display:none;">
                            <label>To search for regulatory motifs in the promoters of the genes please supply the following data:</label>
                            <br><br>
                        <div>
                            <div class="form-group row">
                                <div class="col-md-10">
                                    <label for="genome">Upload a full genome <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> file: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info10">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br>
                                    <label class="btn btn-default btn-file">
                                    <input type="file" id="genome" name="genome" onchange="openCIS()">
                                    </label>
                                </div>
                            </div>
                        </div>
                        <br>
                        <div id="CIS" style="display:none;">
                            <label>Select regulatory elements to be searched for in the promoters: <button type="button" class="btn btn-lg" data-toggle="modal" data-target="#info11">
                                            <span class="glyphicon glyphicon-question-sign"></span>
                                        </button></label><br><br>
                            <input type="checkbox" id="PIP-box" name="PIP-box" value="PIP-box">
                            <label for="PIP-box">PIP-box</label>
                            <br>
                            <input type="checkbox" id="hrp-box" name="hrp-box" value="hrp-box">
                            <label for="hrp-box">hrp-box</label>
                            <br>
                            <input type="checkbox" id="mxiE-box" name="mxiE-box" value="mxiE-box">
                            <label for="mxiE-box">mxiE-box</label>
                            <br>
                            <input type="checkbox" id="exs-box" name="exs-box" value="exs-box">
                            <label for="exs-box">exs-box</label>
                            <br>
                            <input type="checkbox" id="tts-box" name="tts-box" value="tts-box">
                            <label for="tts-box">tts-box</label>
                        </div>
                    </div>
                </div>
            <br><br>
            <button type="submit" class="btn btn-success"> <!--g-recaptcha" data-callback="onSubmit" data-sitekey="6LdIEO4UAAAAAG8KoM5Netgx5LgEJnwfZAngTljA">-->
                <span class="glyphicon glyphicon-cloud-upload"></span>&nbsp;
                Submit
            </button>
            </div>              
            <!-- Modals -->
            <div class="modal fade" id="info1" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            The input for Effectidor is a text file with all the Open Reading Frames (ORFs) sequences of the analyzed genome(s).
                            The file should be in a DNA <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA format</a>.
                            <br>You can find instructions to download this file <a href="instructions.html" target="_blank">here</a>.
                            <br>If you wish to analyzed multiple genomes, the ORFs from each genome should be supplied in separate files and all the files should be compressed together in a <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">ZIP</a> archive.
                            <br>Make sure the names of the fasta files are indicative for you, as these are the names that will appear in the results. For example, 'genome1.fasta','genome2.fasta'.
                            <br>Please make sure all the files packed in your input zip archive are in DNA <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA format</a>, and are <u>not</u> packed in directories - Directories within the zip archive are ignored.
                            <br><br> Note, only protein coding genes should be included in this input. Non-coding RNA genes, such as tRNA, should not be included. If you follow the instructions above, no such genes are expected in your input file. 
                            Alternatively, if such RNA genes do appear in your data and you supply a GFF3 file indicating it (in the advanced options), we will filter them out automatically.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info2" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            This file should be in a DNA <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA format</a>.
                            The records supplied in this file must be copied from the main ORFs file, and will be treated as "known effectors" to train the machine-learning algorithm.
                            <br>When analyzing multiple genomes, you can create such effectors file for each (or some) of the analyzed genomes. In this case, please pack these files in a <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">ZIP</a> archive.
                            Make sure the names of these files are matching to the corresponding ORFs file. For example, if the analyzed ORFs files are 'genome1.fasta','genome2.fasta', the names of the corresponding effectors files should be 'genome1.fna','genome2.fna' (the file extension does not matter).
                            <br>This is an optional (and recommended) input. If you choose not to supply it for a certain genome, the records that show high sequence similarity to our effector dataset (see "Data") will be used as the "known effectors" in this genome.
                            <br>Please note that the predictions are conducted after training the machine-learning algorithm on these "known effectors" of the given genomes, and no exterior effectors' data are used for training. Thus, it is most recommended adding these records to achieve better predictions of potential novel T3Es. This is especially important when not many effectors are known in the bacterial genome you are analyzing.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
             <div class="modal fade" id="info3" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            The previous effectors input (from the analyzed genome(s)) is optional, and if not supplied we build the effectors set based on homology to an established effector dataset.
                            Since you supplied an effectors input, we can either use it as the only positive set, or we can add to it ORFs from your input that share high homology to our effector dataset.
                            Choose "yes" if you want us to search for additional effectors based on homology to use for the training, and "no" if you want your input to be the only data used in the positive set.<br><br>
                            Please note that homology to our effector data is used as a feature in any case.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info4" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            The minimal identity percentage required for a pair of best reciprocal hits to be considered
                            as orthologs (lower values will be filtered out).
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info5" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            The minimal coverage percentage required for a pair of best reciprocal hits to be considered
                            as orthologs (lower values will be filtered out).
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info6" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            In this input you may upload a protein <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> file with type 3 effector records of other bacteria to be used in addition to our effector dataset for homology searches.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info7" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            In this input you may upload host proteome(s).<br>
                            You should upload protein <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> file(s), compressed together in a <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip archive</a>.<br>
                            The files should not be packed inside directories - directories packed in the zip archive will be ignored. Also files with names starting with a dot (.) or an underscore (_) will be ignored.<br>
                            <br><br>
                            If you upload this archive, we will search for sequence similarities between the protein
                            sequences in the compressed FASTA files and the ORFs in your input genome(s).
                            <br>
                            Sequence similarity between bacterial proteins and host proteins is potentially a very
                            important feature for the classification as it captures Eukaryotic domains within the
                            bacterial proteins.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info8" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            In this input you may upload proteome(s) of closely related bacteria that do not encode a T3SS.<br>
                            You should upload protein <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> files, such that each FASTA file contains all the protein sequences of a single bacterium, compressed together in a <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip archive</a>.<br>
                            The files should not be packed inside directories - directories packed in the zip archive will be ignored. Also files with names starting with a dot (.) or an underscore (_) will be ignored.<br>
                            <br><br>
                            These data will be used for sequence similarity search. Each FASTA file within the zip archive, representing a different proteome, will be used for a different search. Sequence similarity to proteins of closely related bacteria without a T3SS is a strong negative feature for T3Es classification.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info9" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            GFF3 files will be used for extraction of genome organization features.<br>
                            If you wish to analyze multiple genomes, such a file should be given for each genome with matching names, for example if the ORFs files are named 'genome1.fasta' and 'genome2.fasta', the corresponding names of the GFF files should be 'genome1.gff3' and 'genome2.gff3' (note that the basename should match, the file extension does not matter).
                            In this case, these files should be compressed in a single <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip archive</a>.<br>
                            The files within the zip archive should not be inside directories - directories packed in the zip archive will be ignored. Also files with names starting with a dot (.) or an underscore (_), or with names not found in the ORFs files will be ignored.<br>
                            
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info10" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            These files will be used to search for cis-regulatory elements in the promoters of the genes, according to your selection.<br>
                            The regulatory elements available for search are PIP-box, hrp-box, mxiE-box, exs-box, and tts-box.<br><br>
                            If you wish to upload more than one file (to analyze multiple genomes), compress them in a single zip archive.<br>
                            In this case, make sure the names of the files are matching to the ORFs and gff files. For example, if the ORFs files are named 'genome1.fasta' and genome2.fasta', the names of the corresponding gff files are 'genome1.gff3' and genome2.gff3', then the corresponding full genome files should be named 'genome1.txt' and 'genome2.txt' (note that the file extension is just an example, and is of no importance).<br>
                            The files should not be packed inside directories - directories packed in the zip archive will be ignored. Also files with names starting with a dot (.) or an underscore (_), or with names not found in the ORFs files will be ignored.<br>
                            
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info11" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            PIP-box is relevant for <i>Xanthomonas</i>, <i>Ralstonia</i>, and <i>Acidovorax</i>.<br>
                            hrp-box is relevant for <i>Pseudomonas syringae</i> and plant pathogens of the family of Enterobacteria.<br>
                            mxiE-box is relevant for <i>Shigella</i>.<br>
                            exs-box is relevant for <i>Pseudomonas aeruginosa</i>.<br>
                            tts-box is relevant for rhizobia.
                            <br><br>
                            Make sure to choose only regulatory elements that are relevant for the bacterium you analyze. Non-relevant regulatory elements may hamper the classification accuracy.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info12" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            This feature can help us find potential T3Es. The disadvantage of this inclusion: longer running time of Effectidor.
                            <br>More information is available in <a href="https://doi.org/10.3389/fpls.2022.1024405" target="_blank">this paper</a>.
                            Please refer to this paper if you use this feature in your analysis.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
            <div class="modal fade" id="info13" role="dialog">
                <div class="modal-dialog">
                    <!-- Modal content-->
                    <div class="modal-content">
                        <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <div class="modal-title"><b>Input specification</b></div>
                        </div>
                        <div class="modal-body">
                        <b>
                            The disadvantage of this inclusion: longer running time of Effectidor.
                            <br>More information about this prediction is available in <a href="https://www.nature.com/articles/s41587-021-01156-3" target="_blank">this paper</a>.
                        </b>
                        </div>
                    </div>
                </div>
            </div>
        </form>
        </div>
        </div>
    <nav role="navigation" class="navbar-inverse navbar-fixed-bottom ng-binding" style="color: #cccccc; text-align: center; font-family: Calibri;">
        <h4>
                Research Site | <a href="https://www.tau.ac.il/~talp/" target="_blank">Pupko Group</a>
        </h4>
    </nav>
</body>
</html>
    